### PR TITLE
dev: test emit_events

### PR DESCRIPTION
**Pull request by DiegoB1911**
This PR implements the unit test for the emit_events function.

## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?
The emit_events function doesn't have any unit tests.

Resolves: #952 

## What is the new behavior?

-  The test_emit_events function generates some events and adds them to the state before calling the emit_events function. This allows the events to be asserted using the spy.assert_emitted function.
-  I created a MockEvent struct to implement the event trait, as the append_keys_and_data function is required for the assert_emitted function to work. Without that implementation, attempts to assert the events would fail because the expected events in the comparison were empty.
- I added a helper function to serialize mock events to match the structure of the emitted events, so they can be properly asserted.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/1003)

### CODE REVIEW

**Comment by enitrat** [comment]
Hi @DiegoB1911, thank you it is the right approach overall.

However, here are some indications on how to make it "better":

1. I am not sure to understand what the `MockEvent` struct is for. You should be able to assert the emission like this:

```rust
        let contract_events = EventsFilterBuilderTrait::from_events(@spy.get_events())
            .with_contract_address(test_address())
            .build();

        for event in evm_events {
            let starknet_event = EventIntoStarknetEvent::into(event);
            contract_events.assert_emitted(@starknet_event);
        };
 ```
 
 2. your `serialize_mock_events` is a good idea, as when we check for events emitted, they are in serialized form. However, a more idiiomatic way would be to define an implementation of the Into trait:
 
 ```rust
     impl EventIntoStarknetEvent of Into<Event, StarknetEvent> {
        fn into(self: Event) -> StarknetEvent {
            let mut serialized_keys = array![];
            let mut serialized_data = array![];
            Serde::<Array<u256>>::serialize(@self.keys, ref serialized_keys);
            Serde::<Array<u8>>::serialize(@self.data, ref serialized_data);
            StarknetEvent { keys: serialized_keys, data: serialized_data }
        }
    } 
```    

(see usage in example above)

Here is a git diff of the changes I would make. As you see the logic is roughly the same.

I also modified a file under snforge_utils to avoid passing ownership when using `assert_emitted`.


```suggestion
diff --git a/crates/evm/src/backend/starknet_backend.cairo b/crates/evm/src/backend/starknet_backend.cairo
index 6d47a9e..78e0cf8 100644
--- a/crates/evm/src/backend/starknet_backend.cairo
+++ b/crates/evm/src/backend/starknet_backend.cairo
@@ -254,9 +254,12 @@ mod tests {
     };
     use crate::test_utils::{evm_address};
     use snforge_std::{
-        test_address, start_mock_call, get_class_hash, spy_events, EventSpyAssertionsTrait
+        test_address, start_mock_call, get_class_hash, spy_events, EventSpyTrait,
+        Event as StarknetEvent
+    };
+    use snforge_utils::snforge_utils::{
+        assert_not_called, assert_called, EventsFilterBuilderTrait, ContractEventsTrait
     };
-    use snforge_utils::snforge_utils::{assert_not_called, assert_called};
     use super::{commit_storage, emit_events};
     use utils::helpers::compute_starknet_address;
     use utils::traits::bytes::U8SpanExTrait;
@@ -276,24 +279,13 @@ mod tests {
         }
     }

-    // Helper function to serialize events into mock events
-    fn serialize_mock_events(
-        events: Array<Event>, ref mock_events: Array<(ContractAddress, MockEvent)>
-    ) {
-        let contract_address = test_address();
-
-        // Serialize and append each event
-        for event in events {
-            let mut serialized_keys = Default::default();
-            let mut serialized_data = Default::default();
-
-            Serde::<Array<u256>>::serialize(@event.keys, ref serialized_keys);
-            Serde::<Array<u8>>::serialize(@event.data, ref serialized_data);
-
-            mock_events
-                .append(
-                    (contract_address, MockEvent { keys: serialized_keys, data: serialized_data })
-                );
+    impl EventIntoStarknetEvent of Into<Event, StarknetEvent> {
+        fn into(self: Event) -> StarknetEvent {
+            let mut serialized_keys = array![];
+            let mut serialized_data = array![];
+            Serde::<Array<u256>>::serialize(@self.keys, ref serialized_keys);
+            Serde::<Array<u8>>::serialize(@self.data, ref serialized_data);
+            StarknetEvent { keys: serialized_keys, data: serialized_data }
         }
     }

@@ -506,10 +498,9 @@ mod tests {
     #[test]
     fn test_emit_events() {
         let mut state: State = Default::default();
-        let mut mock_events = ArrayTrait::<(ContractAddress, MockEvent)>::new();

         // Prepare events
-        let events = array![
+        let evm_events = array![
             Event { keys: array![], data: array![] }, // Empty event
             Event { keys: array![1.into()], data: array![2, 3] }, // Single key, multiple data
             Event {
@@ -520,11 +511,8 @@ mod tests {
             } // Multiple keys and data
         ];

-        // Serialize events and store in mock_events
-        serialize_mock_events(events.clone(), ref mock_events);
-
         // Add events to state
-        for event in events {
+        for event in evm_events.clone() {
             state.add_event(event);
         };

@@ -532,7 +520,14 @@ mod tests {
         let mut spy = spy_events();
         emit_events(ref state).expect('emit events failed');
         assert!(state.events.is_empty());
-        spy.assert_emitted(@mock_events);
+        let contract_events = EventsFilterBuilderTrait::from_events(@spy.get_events())
+            .with_contract_address(test_address())
+            .build();
+
+        for event in evm_events {
+            let starknet_event = EventIntoStarknetEvent::into(event);
+            contract_events.assert_emitted(@starknet_event);
+        };
     }
 }
 // #[test]
@@ -569,4 +564,3 @@ mod tests {
 // }
 // }

-
diff --git a/crates/snforge_utils/src/lib.cairo b/crates/snforge_utils/src/lib.cairo
index 0588fe6..eba213e 100644
--- a/crates/snforge_utils/src/lib.cairo
+++ b/crates/snforge_utils/src/lib.cairo
@@ -150,7 +150,7 @@ pub mod snforge_utils {
     /// A wrapper structure on an array of events emitted by a given contract.
     #[derive(Drop, Clone)]
     pub struct ContractEvents {
-        pub events: Array<Event>
+        pub events: Span<Event>
     }

     pub trait EventsFilterTrait {
@@ -246,33 +246,33 @@ pub mod snforge_utils {
                         }
                     };

-            ContractEvents { events: filtered_events }
+            ContractEvents { events: filtered_events.span() }
         }
     }

     pub trait ContractEventsTrait {
         fn assert_emitted<T, impl TEvent: starknet::Event<T>, impl TDrop: Drop<T>>(
-            self: ContractEvents, event: @T
+            self: @ContractEvents, event: @T
         );
         fn assert_not_emitted<T, impl TEvent: starknet::Event<T>, impl TDrop: Drop<T>>(
-            self: ContractEvents, event: @T
+            self: @ContractEvents, event: @T
         );
     }

     impl ContractEventsTraitImpl of ContractEventsTrait {
         fn assert_emitted<T, impl TEvent: starknet::Event<T>, impl TDrop: Drop<T>>(
-            self: ContractEvents, event: @T
+            self: @ContractEvents, event: @T
         ) {
             let mut expected_keys = array![];
             let mut expected_data = array![];
             event.append_keys_and_data(ref expected_keys, ref expected_data);

+            let contract_events = (*self.events);
             let mut found = false;
             for i in 0
-                ..self
-                    .events
+                ..contract_events
                     .len() {
-                        let event = self.events.at(i);
+                        let event = contract_events.at(i);
                         if event.keys == @expected_keys && event.data == @expected_data {
                             found = true;
                             break;
@@ -283,17 +283,17 @@ pub mod snforge_utils {
         }

         fn assert_not_emitted<T, impl TEvent: starknet::Event<T>, impl TDrop: Drop<T>>(
-            self: ContractEvents, event: @T
+            self: @ContractEvents, event: @T
         ) {
             let mut expected_keys = array![];
             let mut expected_data = array![];
             event.append_keys_and_data(ref expected_keys, ref expected_data);

+            let contract_events = (*self.events);
             for i in 0
-                ..self
-                    .events
+                ..contract_events
                     .len() {
-                        let event = self.events.at(i);
+                        let event = contract_events.at(i);
                         assert(
                             event.keys != @expected_keys || event.data != @expected_data,
                             'Unexpected event was emitted'
```

File: crates/evm/src/model.cairo [hunk 1]
```diff
@@ -223,6 +223,27 @@ pub struct Event {
...
+impl EventTraitImpl of starknet::Event<MockEvent> {
+    fn append_keys_and_data(self: @MockEvent, ref keys: Array<felt252>, ref data: Array<felt252>) {
+        keys.append_span(self.keys.span());
+        data.append_span(self.data.span());
+    }
+    fn deserialize(ref keys: Span<felt252>, ref data: Span<felt252>) -> Option<MockEvent> {
+        Option::None
+    }
+}
+
```
Comment (enitrat) [inline crates/evm/src/model.cairo:246]:
```suggestion
```
Should not be required, see main message

File: crates/evm/src/backend/starknet_backend.cairo [hunk 2]
```diff
@@ -479,6 +502,38 @@ mod tests {
...
+            Event {
+                keys: array![4.into(), 5.into()], data: array![6]
+            }, // Multiple keys, single data
+            Event {
+                keys: array![7.into(), 8.into(), 9.into()], data: array![10, 11, 12, 13]
+            } // Multiple keys and data
+        ];
+
+        // Serialize events and store in mock_events
+        serialize_mock_events(events.clone(), ref mock_events);
```
Comment (enitrat) [inline crates/evm/src/backend/starknet_backend.cairo:524]: prefer the implementation of `Into`, see main message (but same underlying idea)

File: crates/evm/src/backend/starknet_backend.cairo [hunk 3]
```diff
@@ -479,6 +502,38 @@ mod tests {
         assert_called(starknet_address, selector!("set_code_hash"));
         assert_called(starknet_address, selector!("set_nonce"));
     }
+
+    #[test]
+    fn test_emit_events() {
+        let mut state: State = Default::default();
+        let mut mock_events = ArrayTrait::<(ContractAddress, MockEvent)>::new();
```
Comment (enitrat) [inline crates/evm/src/backend/starknet_backend.cairo:509]:
```suggestion
        let mut mock_events = ArrayTrait::<(ContractAddress, MockEvent)>::new();
```

no longer required

File: crates/evm/src/backend/starknet_backend.cairo [hunk 4]
```diff
@@ -274,6 +276,27 @@ mod tests {
...
+
+            Serde::<Array<u256>>::serialize(@event.keys, ref serialized_keys);
+            Serde::<Array<u8>>::serialize(@event.data, ref serialized_data);
+
+            mock_events
+                .append(
+                    (contract_address, MockEvent { keys: serialized_keys, data: serialized_data })
+                );
+        }
+    }
```
Comment (enitrat) [inline crates/evm/src/backend/starknet_backend.cairo:298]:
```suggestion
```

no longer required, replace with an impl of Into

**Comment by DiegoB1911** [comment]
@enitrat Thank you for your review and explanation. I hadn’t considered implementing the Into trait in that way, and I also thought about importing snforge_std::Event as you did ("snforge_std::Event as StarknetEvent"), but I wasn’t sure it was possible to rename imported elements like in other languages, which is why I created the mock event. Thanks for showing me that approach. I'm still learning how traits and implementations work, and how Cairo functions in general, so I really appreciate your help. I’ve made the requested changes. Please let me know if there’s anything else I need to adjust.

**Review by enitrat — APPROVED** [review]
lgtm, great job 💪

**Pull request merged**